igcollect/sfp_dom.py

```diff
@@ -0,0 +1,372 @@
+        )
+
+        # TX Power is 16-bit unsigned integer with LSB equal to 0.1uW
+        # We use the a2_offset_to_volts() function because the equivalence
+        # is similar to volts
+        self.metrics['tx_power']['current'] = a2_offset_to_volts(
+            self._data, SFF_A2_TX_PWR
+        )
+        self.metrics['tx_power']['low_warning'] = a2_offset_to_volts(
+            self._data, SFF_A2_TX_PWR_LWARN
+        )
+        self.metrics['tx_power']['high_warning'] = a2_offset_to_volts(
+            self._data, SFF_A2_TX_PWR_HWARN
+        )
+        self.metrics['tx_power']['low_critical'] = a2_offset_to_volts(
+            self._data, SFF_A2_TX_PWR_LALRM
+        )
+        self.metrics['tx_power']['high_critical'] = a2_offset_to_volts(
+            self._data, SFF_A2_TX_PWR_HALRM
+        )
+
+        # RX Power is 16-bit unsigned integer with LSB equal to 0.1uW
+        # We use the a2_offset_to_volts() function because the equivalence
+        # is similar to volts
+        rx_power = a2_offset_to_int(self._data, SFF_A2_RX_PWR) / 10000
+        self.metrics['rx_power']['current'] = a2_offset_to_volts(
+            self._data, SFF_A2_RX_PWR
+        )
+        self.metrics['rx_power']['low_warning'] = a2_offset_to_volts(
+            self._data, SFF_A2_RX_PWR_LWARN
+        )
+        self.metrics['rx_power']['high_warning'] = a2_offset_to_volts(
+            self._data, SFF_A2_RX_PWR_HWARN
+        )
+        self.metrics['rx_power']['low_critical'] = a2_offset_to_volts(
+            self._data, SFF_A2_RX_PWR_LALRM
+        )
+        self.metrics['rx_power']['high_critical'] = a2_offset_to_volts(
+            self._data, SFF_A2_RX_PWR_HALRM
+        )
+
+
+def parse_args():
+    parser = ArgumentParser()
+    parser.add_argument('--prefix', default='sfp_dom')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    physical_interfaces = get_physical_interfaces()
+
+    for iface in physical_interfaces:
+        try:
+            data = ethtool_get_module_data(iface)
+            print_metrics(args.prefix, iface, data.metrics)
+        except (NoSFPException, UnsupportedSFPException) as exc:
+            print(exc, file=sys.stderr)
+            continue
+
+
+def get_physical_interfaces():
+    scn = '/sys/class/net'
+    physical_interfaces = []
+    for dev in listdir(scn):
+        dev_path = join(scn, dev)
+        if not (islink(dev_path) or isdir(dev_path)):
+            # Skip regular files
+            continue
+        # Physical interfaces have device as a symlink to /sys/class/device_id
+        if islink(join(scn, dev, 'device')):
+            physical_interfaces.append(dev)
+
+    return physical_interfaces
+
+
+def a2_offset_to_celsius(data, offset):
+    """
+    Returns a signed float built from a 16-bit signed twos complement value
+    in increments of 1/256 degrees Celsius
+    """
+    signed = ctypes.c_int16((data[SFF_A2_BASE + (offset)] << 8 | data[
```

Comment:
you can call `a2_offset_to_int()` here

Reply:
I ended up changing this to a struct.unpack.